drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1073 @@
+ * to flip the page for our plane using the atomic API.
+ * 
+ * In this simple example we're only using the primary plane, but we could also
+ * be updating other planes in this same atomic commit.
+ */
+
+static void modeset_draw_dev(int fd, struct modeset_dev *dev)
+{
+	struct modeset_buf *buf;
+	unsigned int j, k, off;
+	int ret;
+
+	dev->r = next_color(&dev->r_up, dev->r, 5);
+	dev->g = next_color(&dev->g_up, dev->g, 5);
+	dev->b = next_color(&dev->b_up, dev->b, 5);
+
+	buf = &dev->bufs[dev->front_buf ^ 1];
+	for (j = 0; j < buf->height; ++j) {
+		for (k = 0; k < buf->width; ++k) {
+			off = buf->stride * j + k * 4;
+			*(uint32_t*)&buf->map[off] =
+				     (dev->r << 16) | (dev->g << 8) | dev->b;
+		}
+	}
+
+	ret = modeset_atomic_commit(fd, buf, dev);
+	if (ret) {
+		fprintf(stderr, "atomic commit failed, %d\n", errno);
+	} else {
+		dev->front_buf ^= 1;
+		dev->pflip_pending = true;
+	}
+}
+
+/*
+ * modeset_cleanup() stays the same.
+ */
+
+static void modeset_cleanup(int fd)
+{
+	struct modeset_dev *iter;
+	drmEventContext ev;
+	int ret;
+
+	/* init variables */
+	memset(&ev, 0, sizeof(ev));
+	ev.version = DRM_EVENT_CONTEXT_VERSION;
+	ev.page_flip_handler = modeset_page_flip_event;
+
+	while (modeset_list) {
+		/* remove from global list */
+		iter = modeset_list;
+		modeset_list = iter->next;
+
+		/* if a page-flip is pending, wait for it to complete */
+		iter->cleanup = true;
+		fprintf(stderr, "wait for pending page-flip to complete...\n");
+		while (iter->pflip_pending) {
+			ret = drmHandleEvent(fd, &ev);
+			if (ret)
+				break;
+		}
+
+		/* restore saved CRTC configuration */
+		if (!iter->pflip_pending)
+			drmModeSetCrtc(fd,
+				       iter->saved_crtc->crtc_id,
+				       iter->saved_crtc->buffer_id,
+				       iter->saved_crtc->x,
+				       iter->saved_crtc->y,
+				       &iter->conn_id,
+				       1,
+				       &iter->saved_crtc->mode);
+		drmModeFreeCrtc(iter->saved_crtc);
+
+		/* destroy framebuffers */
+		modeset_destroy_fb(fd, &iter->bufs[1]);
+		modeset_destroy_fb(fd, &iter->bufs[0]);
+
+		/* free allocated memory */
+		free(iter);
+	}
+}
+
+/* 
+ * This example shows how to use the KMS atomic API to handle page-flips.
+ * Besides the fact that we've only used the primary plane, you could extend
+ * the example to use overlay and cursor planes. The same approach could be
+ * used to do a page-flip: all the planes are updated in local objects and
+ * then drmModeAtomicCommit() can be called to perform a page-flip.
+ * 
+ * Just like vsync'ed double-buffering, the atomic API does not not solve all
+ * the problems that can happen and you have to figure out the best
+ * implementation for your use case.
+ *
+ * If you want more code, I can recommend reading the source-code of:
+ *  - plymouth (which uses dumb-buffers like this example; very easy to understand)
+ *  - kmscon (which uses libuterm to do this)
+ *  - wayland (very sophisticated DRM renderer; hard to understand fully as it
```

Comment:
ITYM weston?

Reply:
I think it's also really important to be clear about what your intended use and audience is here. Are you trying to just show people roughly how the atomic API works, or are you trying to show how to build a complete KMS user for primary planes only, or are you intending to integrate multiple planes, or ... ?

Obviously I'm biased here, but I would also link kms-quads here, as an example of how to build a timed repaint loop which supports multiple outputs, both GL and software rendering, etc. I think that's a good example for people to see how to use the atomic API to build a compositor, or something which needs to animate across multiple outputs.

That's pretty inappropriate for this example - which I would suggest should be 'I literally don't know how the atomic API works and would like to understand how to show my first pixel content on screen'. To that extent, I would limit the scope as much as possible in this example, and link to the others for more advanced use, so this example really does just focus on the details of the atomic API itself, and not anything else.

Reply:
I think I agree with Daniel here. This repo was created when there was no DRM documentation whatsoever, and it was mostly about raising awareness that you can draw to the screen with no xorg or other compositor whatsoever.

I still like that idea. Nevertheless, more complex examples are always welcome as well. Just make sure to be aware of your goal and clearly state it up front. Otherwise, we end up with a mix, and every reviewer might come from a different viewpoint and ask for different extensions.

Reply:
Yes, I agree too.

I wonder if it would make things simpler to run with just one connector? Then there is no need to look at `page_flip_handler2`, but disabling all unused connectors and CRTCs must still be in the atomic request.

Reply:
I agree. Modified the text to explicitly tell that is a simple example of how to use the atomic API.